composer/trainer/trainer.py

```diff
@@ -997,6 +993,79 @@ def _train_loop(self) -> None:
                                                                                    event=Event.EPOCH_END):
                 self._checkpoint_saver.save_checkpoint(state=self.state, seed=self._seed, device=self._device)
 
+    def _handle_cuda_oom(self, e: RuntimeError):
+        """Handles CUDA Out of Memory and rescales if using adaptive grad_accum."""
+        if self.adaptive_grad_accum and "CUDA out of memory" in str(e):
+            # Raise runtime error if training 1 sample at a time still resulted in CUDA out of memory
+            if self.state.grad_accum == self.state.batch_num_samples:
+                raise RuntimeError("CUDA out of memory. Train loop failed with an internal microbatch of size 1")
+            else:
+                self.state.grad_accum = min(2 * self.state.grad_accum, self.state.batch_num_samples)
+                self.logger.metric_batch({'trainer/grad_accum': self.state.grad_accum})
+        else:
+            # If not CUDA out of memory, raise exception to user. Note that this
+            # truncates the call stack back only to this newly raised error
+            raise e
+
+    def _compute_metrics(self, train_metrics: MetricCollection):
+        """Compute training metrics. Adaptively change microbatch size if enabled to maximize GPU usage.
+
+        Args:
+            train_metrics (MetricCollection): Existing train metrics
+        """
+        rerun = True
+        while rerun:
+            try:
+                self.state.model.eval()
+                with torch.no_grad():
+                    # Store results so we only update train_metrics if all batches are processed without OOM
+                    results = []
+                    for eval_microbatch in self._train_data_spec.split_batch(self.state.batch, self.state.grad_accum):
+                        # TODO: Detect if self.run_event(Event.AFTER_DATALOADER) changes the training
+                        # data and if so print a warning that metrics may return unexpected results
+                        outputs, targets = self._original_model.validate(eval_microbatch)
+                        results.append((outputs, targets))
+                    for outputs, targets in results:
+                        train_metrics.update(outputs, targets)
+                rerun = False
+            except RuntimeError as e:
+                self._handle_cuda_oom(e)
+
+    def _train_and_compute_loss(self, use_grad_scaling: bool):
```

Comment:
`_do_batch_fwd_bwd_pass` is a kinda awkward name though, so some other suggestions:
- `_do_batch_pass`
- `_batch_autograd_pass`
- `_compute_batch_loss_and_grad`

Reply:
I agree with renaming -- will wait until others chime in before doing it though. I'm more in favor of `_compute_batch_loss_and_grad` because we're also calling `loss.backward()` and this isn't just a forward pass

Reply:
Prefer to keep the names shorter if possible. How about:
`_train_and_compute_loss` -> `_train_batch`
`_train_batch` -> `_train_batch_inner`
`_train_batch_inner` -> `_train_microbatches`

Reply:
I used Hanlin's, except I did `_train_batch -> _train_batch_with_context`. I also decomposed `_train_microbatches` more and added `_train_microbatch`